gt/main.c

```diff
@@ -244,13 +247,207 @@ print_ip_err_msg(struct gt_packet_headers *pkt_info)
+		ret = hold_arp(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+	} else {
+		ret = hold_nd(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+	}
+
+	if (ret < 0)
+		clear_ether_cache(eth_cache);
+
+	return ret;
+}
 
 static int
-decap_and_fill_eth(struct rte_mbuf *m,
-	struct gt_config *gt_conf, struct gt_packet_headers *pkt_info)
+drop_cache_entry_randomly(struct neighbor_hash_table *neigh, uint16_t ip_ver)
+{
+	int ret;
+	uint32_t entry_id = rte_rand() % neigh->tbl_size;
+	struct ether_cache *eth_cache;
+	uint32_t entry_start_idx = entry_id;
+
+	while (true) {
+		eth_cache = &neigh->cache_tbl[entry_id];
+		if (rte_atomic32_read(&eth_cache->ref_cnt) == 0) {
+			entry_id = (entry_id + 1) % neigh->tbl_size;
+			eth_cache = NULL;
+		} else
+			break;
+
+		if (entry_start_idx == entry_id)
+			break;
+	}
+
+	if (eth_cache == NULL)
+		return -1;
+
+	if (ip_ver == ETHER_TYPE_IPv4) {
+		ret = put_arp(&eth_cache->ip_addr.ip.v4, rte_lcore_id());
+		if (ret < 0)
+			return ret;
+
+		ret = rte_hash_del_key(neigh->hash_table,
+			&eth_cache->ip_addr.ip.v4);
+		if (ret < 0) {
+			RTE_LOG(CRIT, GATEKEEPER,
+				"gt: failed to delete an Ethernet cache entry from the IPv4 neighbor table at %s, we are not trying to recover from this failure!",
+				__func__);
+		}
+		return ret;
+	}
+
+	if (likely(ip_ver == ETHER_TYPE_IPv6)) {
+		ret = put_nd(&eth_cache->ip_addr.ip.v6, rte_lcore_id());
+		if (ret < 0)
+			return ret;
+
+		ret = rte_hash_del_key(neigh->hash_table,
+			&eth_cache->ip_addr.ip.v6);
+		if (ret < 0) {
+			RTE_LOG(CRIT, GATEKEEPER,
+				"gt: failed to delete an Ethernet cache entry from the IPv6 neighbor table at %s, we are not trying to recover from this failure!",
+				__func__);
+		}
+		return ret;
+	}
+
+	return -1;
+}
+
+static struct ether_cache *
+get_new_ether_cache(struct neighbor_hash_table *neigh)
+{
+	int i;
+	struct ether_cache *eth_cache = NULL;
+
+	for (i = 0; i < neigh->tbl_size; i++) {
+		if (rte_atomic32_read(&neigh->cache_tbl[i].ref_cnt) == 0) {
+			eth_cache = &neigh->cache_tbl[i];
+			break;
+		}
+	}
+
+	return eth_cache;
+}
+
+static struct ether_cache *
+gt_neigh_get_ether_cache(struct neighbor_hash_table *neigh,
+	uint16_t inner_ip_ver, void *ip_dst, struct gatekeeper_if *iface)
+{
+	struct ether_cache *eth_cache;
+	struct lls_config *lls_conf;
+
+	int ret = rte_hash_lookup(neigh->hash_table, ip_dst);
+	if (ret >= 0)
+		return &neigh->cache_tbl[ret];
+
+	lls_conf = get_lls_conf();
+	if (inner_ip_ver == ETHER_TYPE_IPv4) {
+		if (!lls_conf->arp_cache.ip_in_subnet(iface, ip_dst))
+			return NULL;
+	} else if (likely(inner_ip_ver == ETHER_TYPE_IPv6)) {
+		if (!lls_conf->nd_cache.ip_in_subnet(iface, ip_dst))
+			return NULL;
+	} else
+		return NULL;
+
+	eth_cache = get_new_ether_cache(neigh);
+	if (eth_cache == NULL) {
+		ret = drop_cache_entry_randomly(neigh, inner_ip_ver);
+		if (ret < 0)
+			return NULL;
+
+		eth_cache = get_new_ether_cache(neigh);
+		if (eth_cache == NULL)
+			return NULL;
```

Comment:
Giving a warning here before returning `NULL` would help network administrators to identify that the cache is overflowing.